monero/wallet_client.go

```diff
@@ -221,49 +250,211 @@ func (c *walletClient) WaitForReceipt(req *WaitForReceiptRequest) (*wallet.Trans
 	accountIdx uint64,
 	amount *coins.PiconeroAmount,
-) (*wallet.TransferResponse, error) {
+	numConfirmations uint64,
+) (*wallet.Transfer, error) {
 	amt, err := amount.Uint64()
 	if err != nil {
 		return nil, err
 	}
-	return c.wRPC.Transfer(&wallet.TransferRequest{
+	amountStr := amount.AsMoneroString()
+	log.Infof("Transferring %s XMR to %s", amountStr, to)
+	reqResp, err := c.wRPC.Transfer(&wallet.TransferRequest{
 		Destinations: []wallet.Destination{{
 			Amount:  amt,
 			Address: string(to),
 		}},
 		AccountIndex: accountIdx,
 	})
+	if err != nil {
+		log.Warnf("Transfer of %s XMR failed: %s", amountStr, err)
+		return nil, fmt.Errorf("transfer failed: %w", err)
+	}
+	log.Infof("Transfer of %s XMR initiated, TXID=%s", amountStr, reqResp.TxHash)
+	transfer, err := c.waitForReceipt(&waitForReceiptRequest{
+		Ctx:              ctx,
+		TxID:             reqResp.TxHash,
+		NumConfirmations: numConfirmations,
+		AccountIdx:       accountIdx,
+	})
+	log.Warnf("Transfer TXID=%s failed to get receipt: %s", reqResp.TxHash, err)
+	if err != nil {
+		return nil, fmt.Errorf("monero TXID=%s receipt failure: %w", reqResp.TxHash, err)
+	}
+	log.Infof("Transfer TXID=%s succeeded with %d confirmations and fee %s XMR",
+		transfer.TxID,
+		transfer.Confirmations,
+		coins.FmtPiconeroAmtAsXMR(transfer.Fee),
+	)
+	return transfer, nil
 }
 
-func (c *walletClient) SweepAll(to mcrypto.Address, accountIdx uint64) (*wallet.SweepAllResponse, error) {
-	return c.wRPC.SweepAll(&wallet.SweepAllRequest{
+func (c *walletClient) SweepAll(
+	ctx context.Context,
+	to mcrypto.Address,
+	accountIdx uint64,
+	numConfirmations uint64,
+) ([]*wallet.Transfer, error) {
+	addrResp, err := c.GetAddress(accountIdx)
+	if err != nil {
+		return nil, fmt.Errorf("sweep operation failed to get address: %w", err)
+	}
+	from := addrResp.Address
+	balance, err := c.GetBalance(accountIdx)
+	if err != nil {
+		return nil, fmt.Errorf("sweep operation failed to get balance: %w", err)
+	}
+	log.Infof("Starting sweep of %s XMR from %s to %s", coins.FmtPiconeroAmtAsXMR(balance.Balance), from, to)
+	if balance.BlocksToUnlock > 0 {
+		log.Infof("Sweep operation waiting %d blocks for balance to fully unlock", balance.BlocksToUnlock)
+		if _, err = WaitForBlocks(ctx, c, int(balance.BlocksToUnlock)); err != nil {
+			log.Warnf("Sweep operation from %s failed waiting to unlock balance", from)
+			return nil, fmt.Errorf("sweep operation failed waiting to unlock balance: %w", err)
+		}
+	}
+	if balance.Balance == 0 {
+		log.Warnf("Sweep operation from %s failed, no balance to sweep", from)
+		return nil, errors.New("no balance to sweep")
+	}
+	reqResp, err := c.wRPC.SweepAll(&wallet.SweepAllRequest{
 		AccountIndex: accountIdx,
 		Address:      string(to),
 	})
+	if err != nil {
+		log.Warnf("Sweep transaction(s) from %s failed, %s", from, err)
+		return nil, fmt.Errorf("sweep_all failed: %w", err)
+	}
+	log.Infof("Sweep transaction started, TX IDs: %s", strings.Join(reqResp.TxHashList, ", "))
```

Comment:
didn't know SweepAll could return mutliple txs, any idea what the diff txs are doing?

Reply:
If there are too many outputs, it can take more than one transaction to sweep them all. We need to handle it, but it is unlikely that there will ever be more than one.